### ISSUE TITLE
[AMDGPU] SIFoldOperands can create V_FMA_F32_e64 instruction that violates constant bus restriction

### ISSUE THREAD

**Issue by jayfoad** [llvm/llvm-project#71685]
Test case: [r.txt](https://github.com/llvm/llvm-project/files/13297858/r.txt)

```
$ llc -march=amdgcn -mcpu=gfx1100 -x mir r.txt -run-pass si-fold-operands -verify-machineinstrs

# After SI Fold Operands
# Machine code for function fma_sgpr_sgpr_use: IsSSA, NoPHIs

bb.0:
  %0:sgpr_32 = IMPLICIT_DEF
  %1:sgpr_32 = S_MOV_B32 1234567
  %3:vgpr_32 = V_FMA_F32_e64 0, %0:sgpr_32, 0, 1234567, 0, %1:sgpr_32, 0, 0, implicit $mode, implicit $exec
  S_ENDPGM 0, implicit %3:vgpr_32

# End machine code for function fma_sgpr_sgpr_use.

*** Bad machine code: VOP* instruction violates constant bus restriction ***
- function:    fma_sgpr_sgpr_use
- basic block: %bb.0  (0x194d4d0)
- instruction: %3:vgpr_32 = V_FMA_F32_e64 0, %0:sgpr_32, 0, 1234567, 0, %1:sgpr_32, 0, 0, implicit $mode, implicit $exec
LLVM ERROR: Found 1 machine code errors.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
```

**Comment by jayfoad** [comment]
Starting from this instruction: `%3:vgpr_32 = V_FMAC_F32_e64 0, %0, 0, %1, 0, %2, 0, 0, implicit $mode, implicit $exec`

First we try folding `%2:vgpr_32 = COPY %1` into `%2` which is legal: it would create `%3:vgpr_32 = V_FMAC_F32_e64 0, %0, 0, %1, 0, %1, 0, 0, implicit $mode, implicit $exec` which has three sgpr operands `%0`, `%1` and `%1`, but only two _different_ sgprs so it still fits within the GFX11 limit of two. Note that the fold is not performed yet, it is just added to the fold list.

Then we try folding `%1:sgpr_32 = S_MOV_B32 1234567` which is legal: it would create `%3:vgpr_32 = V_FMAC_F32_e64 0, %0, 0, 1234567, 0, %2, 0, 0, implicit $mode, implicit $exec` which has one literal and one sgpr operand.

Unfortunately when we apply both of these folds, we end up with `%3:vgpr_32 = V_FMAC_F32_e64 0, %0, 0, 1234567, 0, %1, 0, 0, implicit $mode, implicit $exec` which has two different sgprs and one literal, which exceeds the limit.

**Comment by jayfoad** [comment]
Fixed by #71710.
